libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,354 @@
+  let n = base64UrlEncode(nArray)
+  let e = base64UrlEncode(eArray)
+
+  let newNonce = await self.newNonce()
+  var header = %*{"alg": Alg, "typ": "JWT", "nonce": newNonce, "url": url}
+  if needsJwk:
+    header["jwk"] = %*{"kty": "RSA", "n": n, "e": e}
+  else:
+    # this get will always succeed due to check in beginning of proc
+    header["kid"] = %*(kid.get())
+  return header
+
+proc signedAcmeRequest(
+    self: ref ACMEApi,
+    url: string,
+    payload: auto,
+    key: KeyPair,
+    needsJwk: bool = false,
+    kid: Opt[Kid] = Opt.none(Kid),
+): Future[ACMEResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  try:
+    let acmeHeader = await self.acmeHeader(url, key, needsJwk, kid)
+    var token = toJWT(%*{"header": acmeHeader, "claims": payload})
+    let derPrivKey = key.seckey.rsakey.getBytes.get
+    let pemPrivKey: string = pemEncode(derPrivKey, "PRIVATE KEY")
+    token.sign(pemPrivKey)
+    let rawResponse = await HttpClientRequestRef
+    .post(
+      self.session,
+      url,
+      body = $token.toFlattenedJson(),
+      headers = [("Content-Type", "application/jose+json")],
+    )
+    .get()
+    .send()
+    return ACMEResponse(
+      body: (await rawResponse.getResponseBody()), headers: rawResponse.headers
+    )
+  except CatchableError as exc:
+    raise newException(ACMEError, "Failed to create JWT", exc)
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to send HTTP request to the ACME server", exc)
+
+proc acmeRegister*(
+    self: ref ACMEApi, key: KeyPair
+): Future[ACMERegisterResponse] {.
+    async: (raises: [ACMEError, SerializationError, CancelledError])
+.} =
+  let payload = %*{"termsOfServiceAgreed": true}
+  Json.decode(
+    (
+      await self.signedAcmeRequest(
+        self.directory.newAccount, payload, key, needsJwk = true
+      )
+    ).body,
+    ACMERegisterResponse,
+  )
+
+proc requestChallenge*(
+    self: ref ACMEApi, domains: seq[string], key: KeyPair, kid = Kid
+): Future[ACMEChallengeResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  # request challenge from ACME server
+  var identifiers: seq[JsonNode] = domains.mapIt(%*{"type": "dns", "value": it})
+  let orderPayload = %*{"identifiers": identifiers}
+
+  var challengeResponse = Json.decode(
+    (
+      await self.signedAcmeRequest(
+        self.directory.newOrder, orderPayload, key, kid = Opt.some(kid)
+      )
+    ).body,
+    ACMEChallengeResponse,
+  )
+
+  challengeResponse.orderURL = challengeResponse.headers.getString("location")
+  if challengeResponse.orderURL == "":
+    raise newException(ACMEError, "'location' header not found in ACME response")
+
+  # get challenges
+  let authorizationsResponse =
+    try:
+      Json.decode(
+        (
+          await HttpClientRequestRef
+          .get(self.session, challengeResponse.authzURL)
+          .get()
+          .send()
+          .getResponseBody()
+        ),
```

Comment:
`getResponseBody()` and in request above we use `.body`, could we unify to better approach?